src/components/Approval/Approval.tsx

```diff
@@ -0,0 +1,168 @@
+import { Button, Divider, Flex, Image, Link, SkeletonCircle } from '@chakra-ui/react'
+import { get } from 'lodash'
+import { useEffect, useRef, useState } from 'react'
+import { CountdownCircleTimer } from 'react-countdown-circle-timer'
+import { useFormContext } from 'react-hook-form'
+import { useHistory, useLocation } from 'react-router-dom'
+import { MiddleEllipsis } from 'components/MiddleEllipsis/MiddleEllipsis'
+import { Row } from 'components/Row/Row'
+import { SlideTransition } from 'components/SlideTransition'
+import { RawText, Text } from 'components/Text'
+import { useSwapper } from 'components/Trade/hooks/useSwapper/useSwapper'
+import { useWallet } from 'context/WalletProvider/WalletProvider'
+import { useLocaleFormatter } from 'hooks/useLocaleFormatter/useLocaleFormatter'
+import { bn } from 'lib/bignumber/bignumber'
+import { theme } from 'theme/theme'
+
+type ApprovalParams = {
+  ethFiatRate: string
+}
+
+export const Approval = () => {
+  const history = useHistory()
+  const location = useLocation()
+  const {
+    handleSubmit,
+    formState: { isSubmitting }
+  } = useFormContext()
+  const approvalInterval: { current: NodeJS.Timeout | undefined } = useRef()
+  const [approvalTxId, setApprovalTxId] = useState<string>()
+  const { ethFiatRate } = location.state as ApprovalParams
+
+  const {
+    getValues,
+    formState: { errors }
+  } = useFormContext()
+  const { approveInfinite, checkApprovalNeeded, buildQuoteTx } = useSwapper()
+  const {
+    number: { toCrypto, toFiat }
+  } = useLocaleFormatter({ fiatType: 'USD' })
+  const {
+    state: { wallet }
+  } = useWallet()
+  const { quote, sellAsset, fees } = getValues()
+  const fee = fees.chainSpecific.approvalFee
+  const symbol = sellAsset.currency?.symbol
+
+  const approve = async () => {
+    if (wallet) {
```

Comment:
stylistically early return pattern makes this less of a christmas tree and more readable